replication-api-impl/src/main/java/com/connexta/replication/api/impl/queue/memory/MemorySiteQueue.java

```diff
@@ -0,0 +1,531 @@
+      MemoryQueueBroker broker, String site, int capacity, MeterRegistry registry) {
+    final Tag tag = Tag.of("site", site);
+    final Iterable<Tag> tags = Set.of(tag);
+
+    this.broker = broker;
+    this.site = site;
+    this.capacity = capacity;
+    this.registry = registry;
+    this.count = registry.gauge("replication.queue.tasks", tags, new AtomicInteger());
+    // no need to gauge this one since the priority gauges below will report that automatically
+    this.pendingCount = new AtomicInteger();
+    this.activeCount = registry.gauge("replication.queue.active.tasks", tags, new AtomicInteger());
+    for (int i = MemorySiteQueue.MIN_PRIORITY; i <= MemorySiteQueue.MAX_PRIORITY; i++) {
+      final Iterable<Tag> ptags = Set.of(tag, Tag.of("priority", Integer.toString(i)));
+
+      pendings.add(
+          registry.gaugeCollectionSize(
+              "replication.queue.pending.tasks", ptags, new LinkedList<>()));
+      this.pendingTimers.add(registry.timer("replication.queue.pending.timer", ptags));
+    }
+    this.queuedCounter = registry.counter("replication.queue.queued.tasks", tags);
+    this.requeuedCounter = registry.counter("replication.queue.requeued.tasks", tags);
+    this.successCounter = registry.counter("replication.queue.success.tasks", tags);
+    this.failCounter = registry.counter("replication.queue.fail.tasks", tags);
+    this.timer = registry.timer("replication.queue.overall.timer", tags);
+    this.activeTimer = registry.timer("replication.queue.active.timer", tags);
+  }
+
+  @Override
+  public QueueBroker getBroker() {
+    return broker;
+  }
+
+  @Override
+  public String getSite() {
+    return site;
+  }
+
+  @Override
+  public int size() {
+    return count.get();
+  }
+
+  @Override
+  public int pendingSize() {
+    return pendingCount.get();
+  }
+
+  @Override
+  public int activeSize() {
+    return activeCount.get();
+  }
+
+  @Override
+  public int remainingCapacity() {
+    return capacity - size();
+  }
+
+  @Override
+  public void put(TaskInfo info) throws InterruptedException {
+    final MemoryTask task = new MemoryTask(info, this, registry.config().clock());
+    final Deque<MemoryTask> queue = pendings.get(task.getPriority());
+    final int c;
+
+    putLock.lockInterruptibly();
+    try {
+      // Note that count is used in wait guard even though it is not protected by lock. This works
+      // because count can only decrease at this point (all other puts are shut out by lock), and we
+      // (or some other waiting put) are signalled if it ever changes from capacity. Similarly for
+      // all other uses of count in other wait guards.
+      while (remainingCapacity() <= 0) {
+        notFull.await();
+      }
+      queue.addLast(task);
+      if (count.incrementAndGet() < capacity) {
+        notFull.signal();
+      }
+      c = pendingCount.getAndIncrement();
+    } finally {
+      putLock.unlock();
+    }
+    if (c == 0) {
+      signalNotEmpty();
```

Comment:
Again, there is never a guarantee that you will be the one waking up first because a task was pulled. As such, some other thread might have acquired the lock before you and pulled a task  before you could have added yours. As such, there would be 2 places available in the queue and thus why the LinkedBlockingQueue decided to add that check to reduce the number of times you signal other threads.